sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/randomExpressions.scala

```diff
@@ -44,8 +44,8 @@ abstract class RDG extends UnaryExpression with ExpectsInputTypes with Stateful
   }
 
   @transient protected lazy val seed: Long = child match {
-    case Literal(s, IntegerType) => s.asInstanceOf[Int]
-    case Literal(s, LongType) => s.asInstanceOf[Long]
+    case e if e.dataType == IntegerType => e.eval().asInstanceOf[Int]
+    case e if e.dataType == LongType => e.eval().asInstanceOf[Long]
```

Comment:
What happens if this is not `foldable`, @maropu ? I'm wondering if `e.eval()` is safe here.

Reply:
Yea, you're right. I thought the foldable check had already been done somewhere...